// @derockdev/discord-components-react/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscordUnderlined = exports.DiscordTime = exports.DiscordThreadMessage = exports.DiscordThread = exports.DiscordTenorVideo = exports.DiscordSystemMessage = exports.DiscordSpoiler = exports.DiscordReply = exports.DiscordReactions = exports.DiscordReaction = exports.DiscordQuote = exports.DiscordMessages = exports.DiscordMessage = exports.DiscordMention = exports.DiscordItalic = exports.DiscordInvite = exports.DiscordInlineCode = exports.DiscordHeader = exports.DiscordEmbedFooter = exports.DiscordEmbedFields = exports.DiscordEmbedField = exports.DiscordEmbedDescription = exports.DiscordEmbed = exports.DiscordCustomEmoji = exports.DiscordCommand = exports.DiscordCodeBlock = exports.DiscordButton = exports.DiscordBold = exports.DiscordAttachments = exports.DiscordAttachment = exports.DiscordActionRow = void 0;
/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
const react_component_lib_1 = require("./react-component-lib");
const loader_1 = require("@derockdev/discord-components-core/loader");
loader_1.defineCustomElements();
exports.DiscordActionRow = react_component_lib_1.createReactComponent('discord-action-row');
exports.DiscordAttachment = react_component_lib_1.createReactComponent('discord-attachment');
exports.DiscordAttachments = react_component_lib_1.createReactComponent('discord-attachments');
exports.DiscordBold = react_component_lib_1.createReactComponent('discord-bold');
exports.DiscordButton = react_component_lib_1.createReactComponent('discord-button');
exports.DiscordCodeBlock = react_component_lib_1.createReactComponent('discord-code-block');
exports.DiscordCommand = react_component_lib_1.createReactComponent('discord-command');
exports.DiscordCustomEmoji = react_component_lib_1.createReactComponent('discord-custom-emoji');
exports.DiscordEmbed = react_component_lib_1.createReactComponent('discord-embed');
exports.DiscordEmbedDescription = react_component_lib_1.createReactComponent('discord-embed-description');
exports.DiscordEmbedField = react_component_lib_1.createReactComponent('discord-embed-field');
exports.DiscordEmbedFields = react_component_lib_1.createReactComponent('discord-embed-fields');
exports.DiscordEmbedFooter = react_component_lib_1.createReactComponent('discord-embed-footer');
exports.DiscordHeader = react_component_lib_1.createReactComponent('discord-header');
exports.DiscordInlineCode = react_component_lib_1.createReactComponent('discord-inline-code');
exports.DiscordInvite = react_component_lib_1.createReactComponent('discord-invite');
exports.DiscordItalic = react_component_lib_1.createReactComponent('discord-italic');
exports.DiscordMention = react_component_lib_1.createReactComponent('discord-mention');
exports.DiscordMessage = react_component_lib_1.createReactComponent('discord-message');
exports.DiscordMessages = react_component_lib_1.createReactComponent('discord-messages');
exports.DiscordQuote = react_component_lib_1.createReactComponent('discord-quote');
exports.DiscordReaction = react_component_lib_1.createReactComponent('discord-reaction');
exports.DiscordReactions = react_component_lib_1.createReactComponent('discord-reactions');
exports.DiscordReply = react_component_lib_1.createReactComponent('discord-reply');
exports.DiscordSpoiler = react_component_lib_1.createReactComponent('discord-spoiler');
exports.DiscordSystemMessage = react_component_lib_1.createReactComponent('discord-system-message');
exports.DiscordTenorVideo = react_component_lib_1.createReactComponent('discord-tenor-video');
exports.DiscordThread = react_component_lib_1.createReactComponent('discord-thread');
exports.DiscordThreadMessage = react_component_lib_1.createReactComponent('discord-thread-message');
exports.DiscordTime = react_component_lib_1.createReactComponent('discord-time');
exports.DiscordUnderlined = react_component_lib_1.createReactComponent('discord-underlined');
